core/src/main/scala/org/apache/spark/scheduler/Stage.scala

```diff
@@ -68,6 +68,12 @@ private[scheduler] abstract class Stage(
   /** Set of jobs that this stage belongs to. */
   val jobIds = new HashSet[Int]
 
+  /**
+   * Partitions which there is not yet a task succeeded on. Note that for [[ShuffleMapStage]]
+   * pendingPartitions.size() == 0 doesn't mean the stage is available. Because the succeeded
+   * task can be bogus which is out of date and task's epoch is older than corresponding
+   * executor's failed epoch in [[DAGScheduler]].
+   */
```

Comment:
How about:

Partitions the DAGScheduler is waiting on before it tries to mark the stage / job as completed and continue.  Most commonly, this is the set of tasks that are not successful in the active taskset for this stage, but not always.  In particular, when there are multiple attempts for a stage, then this will include late task completions from earlier attempts.  Finally, note that when this is empty, it does not *necessarily* mean that stage is completed -- we have may have lost some of the map output from that stage.  But the DAGScheduler will check for this condition and resubmit the stage if necessary.